src/combinatorics/simple_graph/triangle/tripartite.lean

```diff
@@ -0,0 +1,264 @@
+  vertical, diagonal lines that intersect in an element of `s`. The explicit triangles are
+  edge-disjoint. Accidental triangles correspond to non-trivial corners.
+
+# Main declarations
+
+* `simple_graph.tripartite_from_triangles.graph`: The tripartite simple graph generated by the given
+  triangle indices.
+* `simple_graph.tripartite_from_triangles.explicit_disjoint`: Predicate on triangle indices for the
+  constructed graph to have edge-disjoint explicit triangles.
+* `simple_graph.tripartite_from_triangles.no_accidental`: Predicate on triangle indices for there to
+  be no accidental triangles.
+* `simple_graph.tripartite_from_triangles.locally_linear`: If both predicates hold, the constructed
+  graph is locally linear.
+-/
+
+open finset function sum3
+
+variables {α β γ 𝕜 : Type*} [linear_ordered_field 𝕜] {t : finset (α × β × γ)} {a a' : α} {b b' : β}
+  {c c' : γ} {x : α × β × γ} {ε : 𝕜}
+
+namespace simple_graph
+namespace tripartite_from_triangles
+
+/-- The underlying relation of the tripartite-from-triangles graph. Two vertices are related iff
+there exists a triangle index containing them both. -/
+@[mk_iff] inductive rel (t : finset (α × β × γ)) : α ⊕ β ⊕ γ → α ⊕ β ⊕ γ → Prop
+| in₀₁ ⦃a b c⦄ : (a, b, c) ∈ t → rel (in₀ a) (in₁ b)
+| in₁₀ ⦃a b c⦄ : (a, b, c) ∈ t → rel (in₁ b) (in₀ a)
+| in₀₂ ⦃a b c⦄ : (a, b, c) ∈ t → rel (in₀ a) (in₂ c)
+| in₂₀ ⦃a b c⦄ : (a, b, c) ∈ t → rel (in₂ c) (in₀ a)
+| in₁₂ ⦃a b c⦄ : (a, b, c) ∈ t → rel (in₁ b) (in₂ c)
+| in₂₁ ⦃a b c⦄ : (a, b, c) ∈ t → rel (in₂ c) (in₁ b)
+
+open rel
+
+lemma rel_irrefl : ∀ x, ¬ rel t x x.
+lemma rel_symm : symmetric (rel t) := λ x y h, by cases h; constructor; assumption
+
+/-- The tripartite-from-triangles graph. Two vertices are related iff there exists a triangle index
+containing them both. -/
+def graph (t : finset (α × β × γ)) : simple_graph (α ⊕ β ⊕ γ) := ⟨rel t, rel_symm, rel_irrefl⟩
+
+namespace graph
+
+@[simp] lemma not_in₀₀ : ¬ (graph t).adj (in₀ a) (in₀ a') .
+@[simp] lemma not_in₁₁ : ¬ (graph t).adj (in₁ b) (in₁ b') .
+@[simp] lemma not_in₂₂ : ¬ (graph t).adj (in₂ c) (in₂ c') .
+
+@[simp] lemma in₀₁_iff : (graph t).adj (in₀ a) (in₁ b) ↔ ∃ c, (a, b, c) ∈ t :=
+⟨by { rintro ⟨⟩, exact ⟨_, ‹_›⟩ }, λ ⟨_, h⟩, in₀₁ h⟩
+@[simp] lemma in₁₀_iff : (graph t).adj (in₁ b) (in₀ a) ↔ ∃ c, (a, b, c) ∈ t :=
+⟨by { rintro ⟨⟩, exact ⟨_, ‹_›⟩ }, λ ⟨_, h⟩, in₁₀ h⟩
+@[simp] lemma in₀₂_iff : (graph t).adj (in₀ a) (in₂ c) ↔ ∃ b, (a, b, c) ∈ t :=
+⟨by { rintro ⟨⟩, exact ⟨_, ‹_›⟩ }, λ ⟨_, h⟩, in₀₂ h⟩
+@[simp] lemma in₂₀_iff : (graph t).adj (in₂ c) (in₀ a) ↔ ∃ b, (a, b, c) ∈ t :=
+⟨by { rintro ⟨⟩, exact ⟨_, ‹_›⟩ }, λ ⟨_, h⟩, in₂₀ h⟩
+@[simp] lemma in₁₂_iff : (graph t).adj (in₁ b) (in₂ c) ↔ ∃ a, (a, b, c) ∈ t :=
+⟨by { rintro ⟨⟩, exact ⟨_, ‹_›⟩ }, λ ⟨_, h⟩, in₁₂ h⟩
+@[simp] lemma in₂₁_iff : (graph t).adj (in₂ c) (in₁ b) ↔ ∃ a, (a, b, c) ∈ t :=
+⟨by { rintro ⟨⟩, exact ⟨_, ‹_›⟩ }, λ ⟨_, h⟩, in₂₁ h⟩
```

Comment:
Is it definitely not worth generalizing to n-cliques (and `sigma`) so that you don't need to repeat things 3 times?

Reply:
I have exactly three applications in mind for this API, and they all use triangles (hell, even locally linear graphs are only about triangles). The point of this API is to repeat things 3 times only once (understand, not in the applications). In contrast, using `sigma` will result in an overhead in the applications.